### PR TITLE
sftp: Fix ENOSPC check

### LINKED ISSUE #3336: Write: sftp: "Failure" (SSH_FX_FAILURE) when the remote storage is full
Output of `restic version`
--------------------------
restic 0.12.0 compiled with go1.15.8 on linux/amd64

How did you run restic exactly?
-------------------------------
`restic -r sftp:root@<ip-adress-snipped>:/mnt/backup-docker/ --verbose prune --password-file=/root/.restic/backup-docker.pwd`

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Scaleway root-server via SFTP

Expected behavior
-----------------

cleaning up / prune snapshots so the backup is working again

Actual behavior
---------------

`restic -r sftp:root@<ip-adress-snipped>:/mnt/backup-docker/ --verbose check --password-file=/root/.restic/backup-docker.pwd`
<details>

```
pack f399332c: not referenced in any index
[...]
A lot of not referenced index files
[...]
pack dfe443af: not referenced in any index
1585 additional files were found in the repo, which likely contain duplicate data.
You can run `restic prune` to correct this.
```

</details>

`restic -r sftp:root@<ip-adress-snipped>:/mnt/backup-docker/ --verbose prune --password-file=/root/.restic/backup-docker.pwd`

<details>

```
repository 92246e2a opened successfully, password is correct
counting files in repo
building new index for repo
[8:50] 100.00%  63722 / 63722 packs
incomplete pack file (will be removed): 17a852a13c08e04fd541e43a593bbe94feba8b77768953a405ccedacfc1648bb
incomplete pack file (will be removed): cf8895aa47d5e19072efeb07f57971bdb9ad929dda3af381d50bd1c14900a44c
[...]
A lot of incomplete pack files
[...]
repository contains 63681 packs (1347596 blobs) with 294.325 GiB
processed 1347596 blobs: 105577 duplicate blobs, 87.518 GiB duplicate
load all snapshots
find data that is still in use for 202 snapshots
[3:11] 18.81%  38 / 202 snapshots
[3:13] 19.31%  39 / 202 snapshots
[3:51] 60.89%  123 / 202 snapshots
[4:02] 96.04%  194 / 202 snapshots
[4:04] 100.00%  202 / 202 snapshots
found 934858 of 1347596 data blobs still in use, removing 412738 blobs
will remove 41 invalid files
will delete 35672 packs and rewrite 346 packs, this frees 255.849 GiB
Save(<data/9f6747a697>) returned error, retrying after 598.193345ms: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/9f6747a697>) returned error, retrying after 850.463688ms: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/9f6747a697>) returned error, retrying after 1.648746395s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/9f6747a697>) returned error, retrying after 1.81251255s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/9f6747a697>) returned error, retrying after 3.385001074s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/9f6747a697>) returned error, retrying after 2.114282783s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/9f6747a697>) returned error, retrying after 7.109876912s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/9f6747a697>) returned error, retrying after 5.809616247s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/9f6747a697>) returned error, retrying after 17.617471184s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/9f6747a697>) returned error, retrying after 12.153765673s: Write: sftp: "Failure" (SSH_FX_FAILURE)
sftp: "Failure" (SSH_FX_FAILURE)itten
Write
github.com/restic/restic/internal/backend/sftp.(*SFTP).Save
	github.com/restic/restic/internal/backend/sftp/sftp.go:326
github.com/restic/restic/internal/limiter.rateLimitedBackend.Save
	github.com/restic/restic/internal/limiter/limiter_backend.go:30
github.com/restic/restic/internal/backend.(*RetryBackend).Save.func1
	github.com/restic/restic/internal/backend/backend_retry.go:56
github.com/cenkalti/backoff.RetryNotify
	github.com/cenkalti/backoff/retry.go:37
github.com/restic/restic/internal/backend.(*RetryBackend).retry
	github.com/restic/restic/internal/backend/backend_retry.go:36
github.com/restic/restic/internal/backend.(*RetryBackend).Save
	github.com/restic/restic/internal/backend/backend_retry.go:50
github.com/restic/restic/internal/cache.(*Backend).Save
	github.com/restic/restic/internal/cache/backend.go:54
github.com/restic/restic/internal/repository.(*Repository).savePacker
	github.com/restic/restic/internal/repository/packer_manager.go:106
github.com/restic/restic/internal/repository.(*Repository).SaveAndEncrypt
	github.com/restic/restic/internal/repository/repository.go:270
github.com/restic/restic/internal/repository.(*Repository).SaveBlob
	github.com/restic/restic/internal/repository/repository.go:704
github.com/restic/restic/internal/repository.Repack
	github.com/restic/restic/internal/repository/repack.go:87
main.pruneRepository
	github.com/restic/restic/cmd/restic/cmd_prune.go:278
main.runPrune
	github.com/restic/restic/cmd/restic/cmd_prune.go:85
main.glob..func18
	github.com/restic/restic/cmd/restic/cmd_prune.go:25
github.com/spf13/cobra.(*Command).execute
	github.com/spf13/cobra/command.go:762
github.com/spf13/cobra.(*Command).ExecuteC
	github.com/spf13/cobra/command.go:850
github.com/spf13/cobra.(*Command).Execute
	github.com/spf13/cobra/command.go:800
main.main
	github.com/restic/restic/cmd/restic/main.go:86
runtime.main
	/usr/lib/go-1.12/src/runtime/proc.go:200
runtime.goexit
	/usr/lib/go-1.12/src/runtime/asm_amd64.s:1337

```
</details>

`restic -r sftp:root@<ip-address-snipped>:/mnt/backup-docker/ --verbose backup /var/lib/docker/ --password-file=/root/.restic/backup-docker.pwd
`
<details>

```
open repository
repository 92246e2a opened successfully, password is correct
lock repository
load index files
using parent snapshot f1a93c2b
start scan on [/var/lib/docker/]
start backup on [/var/lib/docker/]
Save(<data/f5e7222fc9>) returned error, retrying after 437.171317ms: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f73a49bbe2>) returned error, retrying after 423.030259ms: Write: sftp: "Failure" (SSH_FX_FAILURE)eb6e00c8ca55001e77eb36ffc99d5791a179e3cf780aa1c7-json.log
Save(<data/e8ea901191>) returned error, retrying after 412.66944ms: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f5e7222fc9>) returned error, retrying after 729.638889ms: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/e8ea901191>) returned error, retrying after 554.948931ms: Write: sftp: "Failure" (SSH_FX_FAILURE)eb6e00c8ca55001e77eb36ffc99d5791a179e3cf780aa1c7-json.log
Save(<data/f73a49bbe2>) returned error, retrying after 1.091541757s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/e8ea901191>) returned error, retrying after 1.512941144s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f5e7222fc9>) returned error, retrying after 725.284281ms: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f73a49bbe2>) returned error, retrying after 1.546249696s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f5e7222fc9>) returned error, retrying after 2.217300829s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/152265f020>) returned error, retrying after 442.45257ms: Write: sftp: "Failure" (SSH_FX_FAILURE)937b15c2063eb34b23702c1f875e4b90d15e8ceb2aa8f60f1-json.log
Save(<data/e8ea901191>) returned error, retrying after 2.492482153s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/152265f020>) returned error, retrying after 752.170907ms: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f73a49bbe2>) returned error, retrying after 1.567900755s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/152265f020>) returned error, retrying after 866.326182ms: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f5e7222fc9>) returned error, retrying after 2.74188537s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/152265f020>) returned error, retrying after 1.140151404s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f73a49bbe2>) returned error, retrying after 3.288696345s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/e8ea901191>) returned error, retrying after 2.406437136s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/152265f020>) returned error, retrying after 1.664638098s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f5e7222fc9>) returned error, retrying after 2.51257952s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/152265f020>) returned error, retrying after 4.584168338s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/e8ea901191>) returned error, retrying after 4.291992227s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f73a49bbe2>) returned error, retrying after 4.867318033s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f5e7222fc9>) returned error, retrying after 5.731181313s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/e8ea901191>) returned error, retrying after 3.61222373s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/152265f020>) returned error, retrying after 6.138326764s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f73a49bbe2>) returned error, retrying after 4.545179123s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f5e7222fc9>) returned error, retrying after 7.009085549s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/e8ea901191>) returned error, retrying after 10.558950745s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f73a49bbe2>) returned error, retrying after 7.730524827s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/152265f020>) returned error, retrying after 10.06168457s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f5e7222fc9>) returned error, retrying after 17.136020774s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f73a49bbe2>) returned error, retrying after 18.862930378s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/e8ea901191>) returned error, retrying after 11.0497682s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/152265f020>) returned error, retrying after 11.976267608s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/e8ea901191>) returned error, retrying after 20.965183681s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f5e7222fc9>) returned error, retrying after 19.151448088s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/152265f020>) returned error, retrying after 17.99834434s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Save(<data/f73a49bbe2>) returned error, retrying after 18.883270334s: Write: sftp: "Failure" (SSH_FX_FAILURE)
Fatal: unable to save snapshot: Write: sftp: "Failure" (SSH_FX_FAILURE)

```
</details>

Steps to reproduce the behavior
-------------------------------

unfortunately i can't provide any useful information :(

Do you have any idea what may have caused this?
-----------------------------------------------
no :(

Do you have an idea how to solve the issue?
-------------------------------------------
and an aditional "no" :)

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
I really, really love restic and its design principles.
Its an great software which helps me a lot. <3

### PR DESCRIPTION
**Pull request by greatroar**
What does this PR change? What problem does it solve?
-----------------------------------------------------

There were two bugs in the "no space left on device" check in the SFTP backend. With this patch, we check for space that is not reserved for the root user on the remote, and the check is no longer in a defer block because that wouldn't fire. Some change in the surrounding code may have led the deferred function to capture the wrong err variable.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Fixes #3336. Tested on Linux by backing up to an 8MiB ext2 image that has 50% space reserved for root.

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

File: internal/backend/sftp/sftp.go [hunk 1]
```diff
@@ -354,14 +354,13 @@ func (r *SFTP) Save(ctx context.Context, h restic.Handle, rd restic.RewindReader
 			debug.Log("sftp: failed to remove broken file %v: %v",
 				f.Name(), rmErr)
 		}
-
-		err = r.checkNoSpace(dirname, rd.Length(), err)
```
Comment (MichaelEischer) [inline internal/backend/sftp/sftp.go:358]: Overwriting `err` probably has no effect as it isn't a named return parameter. That is, the code currently only overwrites the local variable `err` but has no effect on the returned value.

**Review by MichaelEischer — APPROVED** [review]
LGTM. Could you add a short changelog? The broken check was already introduced as part of restic 0.12.1.

File: changelog/unreleased/issue-3336 [hunk 2]
```diff
@@ -0,0 +1,12 @@
+Enhancement: SFTP backend now checks for disk space
```
Comment (greatroar) [inline changelog/unreleased/issue-3336:1]:
```suggestion
Bugfix: SFTP backend now checks for disk space
```

**Comment by MichaelEischer** [comment]
Thanks!

**Pull request merged**